src/braket/experimental/autoqasm/program/program.py

```diff
@@ -123,6 +125,39 @@ def with_calibrations(self, gate_calibrations: Union[Callable, list[Callable]])
         combined_oqpy_program += self._oqpy_program
         return Program(combined_oqpy_program, has_pulse_control=True)
 
+    def make_bound_program(self, param_values: dict[str, float]) -> Program:
+        """Binds FreeParameters based upon their name and values passed in. If parameters
+        share the same name, all the parameters of that name will be set to the mapped value.
+
+        Args:
+            param_values (dict[str, float]): A mapping of FreeParameter names
+                to a value to assign to them.
+        Returns:
+            Program: Returns a program with all present parameters fixed to their respective
+            values.
+        """
+        # Copy the program so that we don't modify the original program
+        oqpy_program_copy = copy.deepcopy(self._oqpy_program)
+        # Break early if all the parameter assignments have been processed
+        params_to_process = set(param_values.keys())
+
+        # Input declarations only occur at the top level scope
+        state = oqpy_program_copy.stack[0]
+        for i in range(len(state.body)):
+            inst = state.body[i]
+            if isinstance(inst, ast.IODeclaration) and inst.identifier.name in param_values:
+                name = inst.identifier.name
+                target = oqpy_program_copy.declared_vars[name]
+                target.init_expression = param_values[name]
+                new_inst = oqpy.Program().declare(target).stack[0].body[0]
+                state.body[i] = new_inst
```

Comment:
Would this work?
```
state.body[i] = target.make_declaration_statement(bound_oqpy_program)
```

Reply:
That does seem to be a better implementation. For some unfathomable reason, it changes the output of an `np.pi` assignment to:
```
float[64] beta = 3.141592653589793;
```
whereas the current solution results in:
```
float[64] beta = pi;
```
There must be a reason for this, might be worth looking into some more.